core/services/ocr2/plugins/ocr2keeper/evm21/block_subscriber.go

```diff
@@ -0,0 +1,256 @@
+package evm
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/common"
+	ocr2keepers "github.com/smartcontractkit/ocr2keepers/pkg"
+
+	httypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/headtracker/types"
+	"github.com/smartcontractkit/chainlink/v2/core/chains/evm/logpoller"
+	evmtypes "github.com/smartcontractkit/chainlink/v2/core/chains/evm/types"
+	"github.com/smartcontractkit/chainlink/v2/core/logger"
+	"github.com/smartcontractkit/chainlink/v2/core/services/pg"
+	"github.com/smartcontractkit/chainlink/v2/core/utils"
+)
+
+const (
+	// PollLogInterval is the interval to query log poller
+	PollLogInterval = time.Second
+	// CleanUpInterval is the interval for cleaning up block maps
+	CleanUpInterval = 15 * time.Minute
+	// Separator is the separator for block key
+	Separator = "|"
+	// ChannelSize represents the channel size for head broadcaster
+	ChannelSize = 20
+)
+
+type BlockKey struct {
+	block int64
+	hash  common.Hash
+}
+
+func (bk *BlockKey) getBlockKey() ocr2keepers.BlockKey {
+	return ocr2keepers.BlockKey(fmt.Sprintf("%d%s%s", bk.block, Separator, bk.hash.Hex()))
+}
+
+type BlockSubscriber struct {
+	sync                  utils.StartStopOnce
+	mu                    sync.RWMutex
+	ctx                   context.Context
+	cancel                context.CancelFunc
+	hb                    httypes.HeadBroadcaster
+	lp                    logpoller.LogPoller
+	headC                 chan BlockKey
+	unsubscribe           func()
+	subscribers           map[int]chan ocr2keepers.BlockHistory
+	blocksFromPoller      map[int64]common.Hash
+	blocksFromBroadcaster map[int64]common.Hash
+	maxSubId              int
+	lastClearedBlock      int64
+	lastSentBlock         int64
+	blockHistorySize      int64
+	lggr                  logger.Logger
+}
+
+func NewBlockSubscriber(hb httypes.HeadBroadcaster, lp logpoller.LogPoller, blockHistorySize int64, lggr logger.Logger) *BlockSubscriber {
+	return &BlockSubscriber{
+		hb:                    hb,
+		lp:                    lp,
+		headC:                 make(chan BlockKey, ChannelSize),
+		subscribers:           map[int]chan ocr2keepers.BlockHistory{},
+		blocksFromPoller:      map[int64]common.Hash{},
+		blocksFromBroadcaster: map[int64]common.Hash{},
+		blockHistorySize:      blockHistorySize,
+		lggr:                  lggr.Named("BlockSubscriber"),
+	}
+}
+
+func (hw *BlockSubscriber) Start(_ context.Context) error {
+	hw.lggr.Info("block subscriber started.")
+	return hw.sync.StartOnce("BlockSubscriber", func() error {
+		hw.mu.Lock()
+		defer hw.mu.Unlock()
+		_, hw.unsubscribe = hw.hb.Subscribe(&headWrapper{headC: hw.headC})
+		hw.ctx, hw.cancel = context.WithCancel(context.Background())
+
+		// poll from head broadcaster channel and push to subscribers
+		{
+			go func(ctx context.Context) {
```

Comment:
Each of these go functions are really large and can be separated out into their own function calls since they are already scoped. This would possibly help in unit testing too.

Reply:
done